Sage/carbon
# fix(select): fix accessible name

Simple select and Filterable select both use a 2-part aria-labelledby that references both the field's label and a hidden element containing the currently-selected  value. This results in the accessible name being different depending on the selected value, which is unexpected to users who will expect that the accessible name is simply the label.

Note that MultiSelect has been unchanged for now, pending further investigation with UX/accessibility of the best way to make this component accessible.

As a result of these changes, Select and FilterableSelect do not automatically get an accessible name provided, so all uses need either a `label`, `aria-label` or `aria-labelledby` to pass automatic accessibility checks. In particular, `pager` has had to be updated to ensure the page size select always has an accessible name even if `showPageSizeLabelBefore` is false.

fix #5945

### Proposed behaviour

When no `aria-label` or `aria-labelledby` is provided, the accessible name of a simple or filterable select should be simply the visual label.



### Current behaviour

The accessible name of a simple or filterable select is the label followed by the currently-selected value (or "Please select" if there is no currently-selected value).



### Checklist




#### QA


### Additional context



### Testing instructions



Test all 3 select variants, with `aria-label`, `aria-labelledby` and neither, and ensure the behaviour is as expected:
- on simple and filterable select, the accessible name should be the visual label if neither `aria-label` or `aria-labelledby` is provided, and the expected value if one of them is
- on multiselect, there should be no change from the previously-released version (the accessible name consists of the visual label followed by all the currently-selected options)

[Note that the "accessible name" is what is announced by a screenreader along with "combobox" and the current value, and can be inspected in the developer tools in Chrome and Firefox.]

Test with a screenreader and ensure the components are still usable and the announcements make sense. (Not needed for MultiSelect other than to ensure there are no serious regressions - accessibility improvements to MultiSelect are currently under discussion.)

Test the Pager component with a screenreader with all combinations of the `showPageSizeLabelBefore` and `showPageSizeLabelAfter` props (with `showPageSizeSelection=true`) to ensure there is an appropriate accessible name for the select in all cases.
